### ISSUE TITLE
should warn/error on forward references during field initialization

### ISSUE THREAD

**Issue by Spongman** [Microsoft/TypeScript#12673]
TypeScript Version:  2.0.10

```ts
class Test
{
    _b = this._a; // undefined, no error/warning
    _a = 3;

    static _B = Test._A; // undefined, no error/warning
    static _A = 3;
    
    method()
    {
        let a = b; // Block-scoped variable 'b' used before its declaration
        let b = 3;
    }
}
```

I'd expect the forward field references to trigger an error (or at least a warning, maybe configurable) similar to function locals.

**Comment by SlurpTheo** [comment]
**TypeScript Version:**  2.1.4
I'm not certain if this is the same issue/bug and I understand how things gets convoluted quickly when thinking about function _declarations_ and how they hoist within the function where they are defined, but I was recently surprised by the behaviour of a function _expression_ accessing a block-scoped "outer/free" variable (e.g., `let x`) before its declaration.

**Code**
```ts
function B() {
    // comment-out the '**A' lines to see the 'used before its declaration' error

    ;(function() {   // **A
        ++x;
    })()             // **A

    let x = 0
}
```

**Expected behavior:**
A compile error both with and without the nested function expression.

**Actual behavior:**
Use of a block-scoped, `let` variable before its initialization.

**Comment by SlurpTheo** [comment]
```node
> (()=>{ (()=>x)(); let x; })()
ReferenceError: x is not defined
    at repl:1:13
    at repl:1:15
    at repl:1:28
    at sigintHandlersWrap (vm.js:22:35)
    at sigintHandlersWrap (vm.js:96:12)
    at ContextifyScript.Script.runInThisContext (vm.js:21:12)
    at REPLServer.defaultEval (repl.js:313:29)
    at bound (domain.js:280:14)
    at REPLServer.runBound [as eval] (domain.js:293:12)
    at REPLServer.<anonymous> (repl.js:513:10)
> (()=>{ let x; (()=>x)(); })()
undefined
> process.version
'v6.9.2'
>
```

**Comment by RyanCavanaugh** [comment]
@SlurpTheo I think you're basically looking for #11498

**Comment by SlurpTheo** [comment]
@RyanCavanaugh I think that is very close to my concern -- yes/thanks!

(My 2nd https://github.com/Microsoft/TypeScript/issues/12673#issuecomment-267387786 from earlier today shows how `--target es5` will generate code that will behave/error differently from code output via `--target es2015`)
